Framework/HistogramData/src/HistogramItem.cpp

```diff
@@ -0,0 +1,106 @@
+#include "MantidHistogramData/HistogramItem.h"
+#include "MantidHistogramData/Histogram.h"
+#include <algorithm>
+#include <utility>
+
+using Mantid::HistogramData::HistogramItem;
+using Mantid::HistogramData::Histogram;
+using Mantid::HistogramData::BinEdges;
+using Mantid::HistogramData::Points;
+
+HistogramItem::HistogramItem(const Histogram &histogram, const size_t index)
+    : m_histogram(histogram), m_index(index) {}
+
+double HistogramItem::counts() const {
+  const auto &yMode = histogramRef().yMode();
+  if (yMode == Histogram::YMode::Counts) {
+    return histogramRef().counts()[m_index];
```

Comment:
I think all these methods should avoid calling the higher-level methods like `counts()` and `binEdges()`, since they return by value, which, while not a copy of the data, takes a copy of the `cow_ptr` (internally a `shared_ptr`). Doing so once per histogram is usually not too expensive, but here it is done once for every bin so I think we must avoid it to see good performance.

That is, work with `x`, y`, etc. from `Histogram` directly.

Reply:
Ah ok. I didn't want to create any extra work for myself, but if there's a likely performance gain then that's fine.